design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,223 @@
+        - name: OperatorDown
+          expr: example_operator_number_of_pods == 0
+          for: 5m
+          annotations:
+            summary: Example operator is down
+            description: Example operator is down for more than 5 minutes.
+            runbook_url: $RUNBOOK_BASE_URL/OperatorDown
+          labels:
+            severity: critical
+```
+
+But components can have a single file only for metrics or for a specific group
+of metrics, for example:
+
+```yaml
+observability:
+   groups:
+      - name: operator_resources
+        common_labels:
+           - name: grouped
+             value: this_is_grouped
+        metrics:
+           - name: example_operator_reconcile_count
+             help: Number of times the operator has executed the reconcile loop
+             type: counter
+             stability: stable
+             labels:
+                - name: another_label
+                  value: only_for_this_metric
+
+      - name: other_group
+        metrics:
+          ...
+```
+
+In the end, all the declarative configuration will be parsed and joined into a
+single `Observability` struct, which will be used to create the metrics, alerts,
+and recording rules. The same idea will be applied to other resources.
+
+It should be possible to:
+
+- create a single file with all groups of metrics, alerts, and recording rules
+
+- create multiple files, each with a specific set of groups. With each group
+containing its metrics, alerts, and recording rules
+
+- create multiple files, each with multiple groups with only one type of
+resource (metrics, alerts, or recording rules). For example, one file with all
+metrics, another with all alerts, and another with all recording rules.
+
+### Utilities
+
+The operator observability package will also have utilities to:
+- Generate the documentation for metrics, alerts, and recording rules
+- Lint metrics and recording rules names
+- Validate alerts configuration
+- Generate resource definitions PrometheusRule, ServiceMonitor, etc
+- Define allow-list, deny-list, and opt-in metrics
+
+## Action Plan
+
+### Phase 1 - Decouple monitoring logic from business logic
+
+1) Move the monitoring logic from the business logic code to a dedicated package
+in each component repository
+
+### Phase 2 - Use a generic operator-observability package
+
+1) Create the operator-observability package with the generic monitoring code
+   and utilities
+
+2) Document how to use the operator-observability package for adding new
+   metrics, alerts, and recording rules
+
+3) Update the monitoring code in each component to use the
+operator-observability package
+
+### Phase 3 - Have a declarative way to define metrics and alerts
+
+1) Update the operator-observability package to have a declarative way to
+define metrics and alerts
+
+2) Update the monitoring code in each component to use the declarative way to
+define metrics and alerts
+
+### Phase 4 - Move existing monitoring code to kubevirt/monitoring
```

Comment:
I think this can bring complexity regarding versions dependency, like we have now with kube-state-metrics and that its important that we keep the metrics and alerts in each sub operators separately.
Please outline pros and cons for this step. Im not convinced.

Reply:
@sradco Hi. Can you please elaborate on the versions dependency issue you had with kube-state-metrics?

Reply:
@enp0s3 +1

We need more details here w.r.t. the cons.
In general, I think it's very valuable in many ways:
1. We can ensure that monitoring code does not change business logic
2. PRs would be much simpler to review and add when the monitoring and kubevirt-core code would be decoupled. The monitoring experts could own their repo including reviewing, approving and be responsible for the monitoring code in an independent way.
3. CI could look entirely different on both Repos. For example, in the monitoring repo we can add monitoring linters and specific monitoring tests.
4. From the Kubevirt perspective: replace the burden of maintaining monitoring logic and instead focus on testing the monitoring package with Kubevirt.
5. Easier to fork and customize
6. A central place to add docs, bug reports, issues for enhancement requests, monitoring discussions, etc

Reply:
> @enp0s3 +1
> 
> We need more details here w.r.t. the cons. In general, I think it's very valuable in many ways:
> 
> 1. We can ensure that monitoring code does not change business logic

Having it in a separate package inside the repo should allow that

> 2. PRs would be much simpler to review and add when the monitoring and kubevirt-core code would be decoupled. The monitoring experts could own their repo including reviewing, approving and be responsible for the monitoring code in an independent way.

Same, if its in a separate package inside the repo

> 3. CI could look entirely different on both Repos. For example, in the monitoring repo we can add monitoring linters and specific monitoring tests.

I don't think there would be a big advantage there. We need to test them together, see that alerts are fires, metrics are propagated, etc. This can actually create duplicated tests in my opinion.

> 4. From the Kubevirt perspective: replace the burden of maintaining monitoring logic and instead focus on testing the monitoring package with Kubevirt.

I dont see how this cant be separated completely, since there are metrics/logs that are gathers on the VMIs that cant be collected from an external operator.
If all the monitoring code is separated to a different package inside the repo, I don't see why it would still be a burden. The package can have its own maintainers list.

> 5. Easier to fork and customize

If you are using only Kubevirt, why would you want a repo that includes monitoring for the other operators?
> 6. A central place to add docs, bug reports, issues for enhancement requests, monitoring discussions, etc

This would be better scoped when its tied to the specific operator.
It would make it harder to track each individual bug/RFE for the different operators and many times the developers that will need to work on it are the the core operator developers and not the monitoring team developers.

Reply:
Im all for an observability library like Joao suggested, but against separating out the monitoring code from each of the operators to a separate operator.